### PR TITLE
babel-parser(ts): Raise recoverable error for abstract interface

### LINKED ISSUE #12768: babel-parser(ts): Weird AST for abstract interface when errorRecovery is true
## Bug Report

- [x] I would like to work on a fix!

**Current behavior**

Weird AST is returned:

<details>
<div>

```json
[
  {
    "type": "ExpressionStatement",
    "expression": {
      "type": "Identifier",
      "name": "abstract"
    }
  },
  {
    "type": "TSInterfaceDeclaration",
    "id": {
      "type": "Identifier",
      "name": "I"
    },
    "body": {
      "type": "TSInterfaceBody",
      "body": []
    }
  }
]
```

</div>
</details>

- [REPL](https://babeljs.io/repl#?browsers=safari%2012&build=&builtIns=false&spec=false&loose=false&code_lz=IYIwzgLgTsDGEAICWA7CBTKAzO6EDEB7QhAbwF8AoIA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=typescript&prettier=false&targets=&version=7.12.14&externalPlugins=)

**Input Code**

```ts
abstract interface Foo {}
```

**Expected behavior**

Syntax error is thrown. `'abstract' modifier can only appear on a class, method, or property declaration`.
And returns the AST like below:

<details>
<div>

```json
{
  "type": "TSInterfaceDeclaration",
  "id": {
    "type": "Identifier",
    "name": "I"
  },
  "body": {
    "type": "TSInterfaceBody",
    "body": []
  },
  "abstract": true
}
```

</div>
</details>

**Babel Configuration (babel.config.js, .babelrc, package.json#babel, cli command, .eslintrc)**
- Filename: `babel.config.js`

```js
const code = `abstract interface Foo {}`;
const result = parse(code, {
  sourceType: "module",
  errorRecovery: true,
  plugins: ["typescript"],
});
```

**Environment**

**Possible Solution**

**Additional context**

from prettier/prettier#10252.
In previous version, babel/parser throw syntax error `SyntaxError: Unexpected token, expected ";"`.

### PR DESCRIPTION
**Pull request by sosukesuzuki**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #12768
| Patch: Bug Fix?          | Y
| Tests Added + Pass?      | Yes
| License                  | MIT

### CODE REVIEW

**Review by nicolo-ribaudo — APPROVED** [review]

**Review by existentialism — APPROVED** [review]

**Review by kaicataldo — APPROVED** [review]

**Comment by nicolo-ribaudo** [comment]
Could you rebase? CI should be fixed now.

File: packages/babel-parser/src/plugins/typescript/index.js [hunk 1]
```diff
@@ -2849,4 +2844,31 @@ export default (superClass: Class<Parser>): Class<Parser> =>
         this.state.inAbstractClass = oldInAbstractClass;
       }
     }
+
+    tsParseAbstractDeclaration(
+      node: any,
+    ): N.ClassDeclaration | N.TsInterfaceDeclaration {
+      const isClass = this.match(tt._class);
+      if (!isClass && !this.isContextual("interface")) {
```
Comment (JLHwung) [inline packages/babel-parser/src/plugins/typescript/index.js:2852]: nit: I prefer to move this branch to the `else` branch of `isClass`:
```
if (isClass) {
  parseClass(...)
} else if (this.isContextual("interface")) {
  throw on abstract interface;
} else {
  throw on unexpected token;
}
```

File: packages/babel-parser/src/plugins/typescript/index.js [hunk 2]
```diff
@@ -2849,4 +2844,31 @@ export default (superClass: Class<Parser>): Class<Parser> =>
...
+          /* optionalId */ false,
+        );
+      } else {
+        // for invalid abstract interface
+        this.raise(
+          node.start,
+          TSErrors.NonClassMethodPropertyHasAbstractModifer,
+        );
+        this.next();
+        return this.tsParseInterfaceDeclaration(
```
Comment (JLHwung) [inline packages/babel-parser/src/plugins/typescript/index.js:2869]: `interface` should not have a proceeding line break:
```ts
interface
Foo {}
```
is invalid. So it does not suffice to conclude that it is an interface declaration when we see a contextual `interface` name.

**Review by JLHwung — APPROVED** [review]
LGTM with some edge cases and code nits.

File: packages/babel-parser/src/plugins/typescript/index.js [hunk 3]
```diff
@@ -2849,4 +2844,36 @@ export default (superClass: Class<Parser>): Class<Parser> =>
...
+          /* isStatement */ true,
+          /* optionalId */ false,
+        );
+      } else if (this.isContextual("interface")) {
+        // for invalid abstract interface
+
+        // To avoid
+        //   abstract interface
+        //   Foo {}
+        if (!this.hasPrecedingLineBreak(this.lookahead())) {
```
Comment (sosukesuzuki) [inline packages/babel-parser/src/plugins/typescript/index.js:2864]: Is there a way to avoid this `lookahead`?

Comment (nicolo-ribaudo) [reply packages/babel-parser/src/plugins/typescript/index.js:2864]: Maybe with a combination of `lineBreak.test` and `this.nextTokenStart`

Comment (sosukesuzuki) [reply packages/babel-parser/src/plugins/typescript/index.js]: It seems to work fine! da8ce69

**Review by JLHwung — APPROVED** [review]

**Review by nicolo-ribaudo — APPROVED** [review]

**Pull request merged**